File: android/PriceFinder/app/src/main/java/enterprises/stimes/pricefinding/scraping/BestBuyScraper.java
package enterprises.stimes.pricefinding.scraping;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;

public class BestBuyScraper extends Scraper {
	
	private String apiKey = Config.bestBuyApiKey;
	private String url = "http://api.bestbuy.com/v1/products";
			//+ "(name=castlevania*)?";//;apiKey=qt6ty86ctjj5f74s8gsnp4yn&format=json";
	
	private String apiKeyParam = "apiKey";
	//private String queryParam = "name";
	
	public BestBuyScraper(){
		super("BestBuy");
	}

	public ArrayList<ScrapingItem> search(String keyword)  {
		this.keywords = keyword;
		try {
			String query = String.format("(%s)?%s=%s", 
				 
			     //queryParam,
			     URLEncoder.encode(getSearchText(), charset),
					apiKeyParam,
				     URLEncoder.encode(this.apiKey, charset) 
			     );
			
			
			String queryUrl = url + query + "&format=json";
			System.out.println(queryUrl);
			JSONObject json = ScrapeUtils.readJsonFromUrl(queryUrl);

			return super.getItemsFromSearch(json);
		} 
		catch (UnsupportedEncodingException e1) {
			e1.printStackTrace();
			return new ArrayList<ScrapingItem>();
		}
		catch(JSONException e){
			e.printStackTrace();
			return new ArrayList<>();
		}
	}
	
	private String getSearchText(){
		String toReturn = "";
		String[] keywrds = keywords.split(" ");
		boolean amp = false;
		for(String keyword : keywrds) {
			if(amp){
				toReturn += "&";
			}
			toReturn += "search=" + keyword;
			
			amp = true;
		}
		return toReturn;
	}
	
}

File: ios/PriceFinder/PriceFinder/PriceFinding/BestBuyScraper.swift
//
//  BestBuyScraper.swift
//  PriceFinder
//
//  Created by jacob stimes on 7/25/16.
//  Copyright © 2016 stimes.enterprises. All rights reserved.
//

import Foundation

/// Queries the Best buy api
class BestBuyScraper : Scraper, ScraperProtocol {
    
    let apiKey = Config.BestBuyApiKey;
    let url = "https://api.bestbuy.com/v1/products";
    let apiKeyParam = "apiKey";
    
    init() {
        super.init(site:"Best Buy");
    }
    
    func search(keywords: String) -> [Item] {
        self.keywords = keywords;
        
        let s = NSCharacterSet.URLQueryAllowedCharacterSet().mutableCopy()
        s.addCharactersInString("+&")
        var query = "(" + getSearchText().stringByAddingPercentEncodingWithAllowedCharacters(s as! NSCharacterSet)! + ")?";
        
        query = query + apiKeyParam + "=" + apiKey.stringByAddingPercentEncodingForRFC3986()!;
        query = query + "&format=json"
        query = query + "&show=thumbnailImage,mobileUrl,name,onSale,regularPrice,salePrice,shortDescription&pageSize=20".stringByAddingPercentEncodingWithAllowedCharacters(s as! NSCharacterSet)!;
        
        let queryUrl = url + query;
        
        let json = self.parseIntoJSON(self.getJSON(queryUrl));
        return getItemsFromJSON(json);
    }
    
    func getItemsFromJSON(results: JSON) -> [Item] {
        var items = [Item]();
        
        let itemsNode = "products";
        
        let itemsJsonArray = results[itemsNode] as! [JSON];
        for itemJson in itemsJsonArray {
            let name = itemJson["name"] as! String;
            let salePrice = itemJson["salePrice"] as! NSNumber;
            let link = itemJson["mobileUrl"] as! String;
            
            var thumbnailLink = "";
            if let tryLink = itemJson["thumbnailImage"] as? String {
                thumbnailLink = tryLink;
            }
            
            items.append(Item(title: name, price: Double(salePrice), vendor: site, link: link, thumbnailLink: thumbnailLink));
        }
        
        return items;
    }
    
    private func getSearchText() -> String {
        var toReturn = "";
        var amp = false;
        
        for keyword in self.keywords!.componentsSeparatedByString(" ") {
            if amp {
                toReturn += "&";
            }
            toReturn += "search=" + keyword;
            
            amp = true;
        }
        
        return toReturn;
    }
}